### PR TITLE
`tpctl debug` command to generate debug commands in .env

### PR DESCRIPTION
**Pull request by ofey404**
## Example
```bash
$ tpctl debug --deploy-id tpctl-hello-test-tpctl-q948q --cluster-namespace tpctl-hello-test-tpctl-cron
generate .env in current directory
Run:
     source .env
to get debug commands.

$ source .env
available commands:
- t_log_case
- t_ls_pod
- t_log_pod {pod-name}
- t_ssh_pod {pod-name}

logs are stored in ./output/

$ t_ls_pod 
NAME                                                     READY   STATUS    RESTARTS   AGE
tpctl-hello-test-tpctl-cron-discovery-7ffd749cb4-zzk9x   1/1     Running   0          4d2h
tpctl-hello-test-tpctl-cron-monitor-6c54965968-gj26w     3/3     Running   0          4d2h
tpctl-hello-test-tpctl-cron-pd-0                         1/1     Running   0          4d2h
tpctl-hello-test-tpctl-cron-tidb-0                       2/2     Running   0          4d2h
tpctl-hello-test-tpctl-cron-tikv-0                       1/1     Running   0          4d2h
tpctl-hello-test-tpctl-cron-tikv-1                       1/1     Running   0          4d2h
tpctl-hello-test-tpctl-cron-tikv-2                       1/1     Running   0          4d2h
tpctl-hello-test-tpctl-cron-tikv-3                       1/1     Running   0          4d2h
tpctl-hello-test-tpctl-cron-tikv-4                       1/1     Running   0          4d2h
```

### CODE REVIEW

File: tipocket-ctl/tpctl/debug.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+import click
+
+HELP_STRING = """
+generate debug environment into .env file.
+
+    Run `source .env` to get commands.
+
+\b
+You can find argument {deploy-id} and {cluster-namespace} in slack notification:
```
Comment (cosven) [inline tipocket-ctl/tpctl/debug.py]: we can get the `cluster-namespace` from the `deploy-id`.

File: tipocket-ctl/tpctl/debug.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+"""
+
+
+def generate_script(deploy_id, cluster_namespace):
+    variables = inspect.cleandoc("""
+    DEPLOY_ID={}
+    CLUSTER_NAMESPACE={}
+    """.format(deploy_id, cluster_namespace))
+
+    functions = inspect.cleandoc("""
```
Comment (cosven) [inline tipocket-ctl/tpctl/debug.py]: We'd better save the script content in a seperate `.sh` file.
It will be more convinient to debug and modify this script.

File: tipocket-ctl/tpctl/debug.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+            kubectl -n $CLUSTER_NAMESPACE exec --stdin --tty $1 -- /bin/sh
+        fi
+    }
+
+    """)
+    return variables + '\n' + functions
+
+
+def print_debug_help():
+    click.echo('generate .env in current directory')
```
Comment (cosven) [inline tipocket-ctl/tpctl/debug.py]: Let's capitalize the first charactor.

File: tipocket-ctl/tpctl/debug.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    click.echo('Run:')
+    click.secho('source .env', fg='green')
+    click.echo('to get debug commands.')
+
+
+@click.command(help=HELP_STRING)
+@click.option('--deploy-id', help='id of test argo workflow')
```
Comment (cosven) [inline tipocket-ctl/tpctl/debug.py]: The `deploy_id` is a required argument. We can use `click.argument`.

File: tipocket-ctl/tpctl/debug.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+    """)
+    return variables + '\n' + functions
+
+
+def print_debug_help():
+    click.echo('generate .env in current directory')
+    click.echo('Run:')
+    click.secho('source .env', fg='green')
+    click.echo('to get debug commands.')
```
Comment (cosven) [inline tipocket-ctl/tpctl/debug.py]: How about put them in one line? `click.style` should do some help.

File: tipocket-ctl/tpctl/debug.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    """
+    deploy_id = params['deploy_id']
+    cluster_namespace = params['cluster_namespace']
+
+    if deploy_id is None or cluster_namespace is None:
+        click.echo('No deploy-id or cluster-namespace specified')
+        click.echo('Refer to slack notification')
+        click.echo('Or run:')
+        click.secho('tpctl debug --help', fg='green')
+        return
```
Comment (cosven) [inline tipocket-ctl/tpctl/debug.py]: This can be omitted when we use `click.argument`.

File: tipocket-ctl/tpctl/debug.py (hunk 2 above)
Comment (ofey404) [reply tipocket-ctl/tpctl/debug.py]: Where should we place the script file?
Under the project root directory or in a separate directory like asset/env.sh?

I prefer a separate directory like asset/ or script/.

File: tipocket-ctl/tpctl/debug.py (hunk 1 above)
Comment (ofey404) [reply tipocket-ctl/tpctl/debug.py]: So we omit the possibility that namespace are specified in {namespace} flag?

Or maybe we should remove this flag and generate every namespace in tpctl.

Comment (cosven) [reply tipocket-ctl/tpctl/debug.py]: We can get the namespace from the argo workflow spec.

For example:
```
argo get {workflow} | grep namespace | xxx
```

-------------

It may be hard to parse the namespace in this way and I think we can set a specific `label` on the argo workflow. For example:
```
workflow['metadata']['labels']['tpctl-resource-namespace'] = xxx
```

File: tipocket-ctl/tpctl/debug.py (hunk 2 above)
Comment (cosven) [reply tipocket-ctl/tpctl/debug.py]: > I prefer a separate directory like asset/ or script/

+1.

I think we can have such layout

```
- tpctl/
  - prepare.py
  - ...
  - scripts/
     - xxx.sh
```

ref: https://stackoverflow.com/questions/6028000/how-to-read-a-static-file-from-inside-a-python-package

File: tipocket-ctl/tpctl/debug.py (hunk 1 above)
Comment (ofey404) [reply tipocket-ctl/tpctl/debug.py]: Good idea.

Comment (ofey404) [reply tipocket-ctl/tpctl/debug.py]: I parsed the output of `argo get -n argo $DEPLOY_ID -o json` with jq.

File: tipocket-ctl/tpctl/scripts/env_raw.sh [hunk 7]
```diff
@@ -0,0 +1,72 @@
+# WARNING: All commands connect to `tidb` container of tidb pods by default.
```
Comment (cosven) [inline tipocket-ctl/tpctl/scripts/env_raw.sh:1]: We should add `scripts/env_raw.sh` as tpctl package data.
https://packaging.python.org/guides/distributing-packages-using-setuptools/#package-data

File: tipocket-ctl/tpctl/scripts/env_raw.sh [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+# 	   As tidb pod has two containers
+OUTPUT_DIR=output
+CASE_POD_ID=$(argo get -n argo $DEPLOY_ID -o json | jq -r '.status.nodes[] | select(.type == "Pod" and .templateName != "notify") | .id')
+CLUSTER_NAMESPACE=$(argo get -n argo $DEPLOY_ID -o json | jq -r '.spec.templates[-1].container.command[2]' | grep -oP '\-namespace=".*?"' | grep -o '".*"' | sed -e 's/^"//' -e 's/"$//')
```
Comment (cosven) [inline tipocket-ctl/tpctl/scripts/env_raw.sh]: Can we shorten this line?

File: tipocket-ctl/tpctl/debug.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+============ Slack Notification ======================
+argo workflow
+tpctl-hello-test-tpctl-q948q
+^^^ {DEPLOY ID} ^^^^^^^^^^^^
+=======================================================
+"""
+
+
+def generate_script(deploy_id):
+    variables = inspect.cleandoc("""
```
Comment (cosven) [inline tipocket-ctl/tpctl/debug.py]: maybe just
```
variables = f'DEPLOY_ID={deploy_id}'
```

**Review by cosven — COMMENTED** [review]
rest LGTM

**Comment by ofey404** [comment]
Now output has become:
```bash
$ tpctl debug tpctl-hello-test-tpctl-q948q
Generate .env in /tmp/tpctl-hello-test-tpctl-q948q
Run to get debug commands:
cd /tmp/tpctl-hello-test-tpctl-q948q
source .env
```

File: tipocket-ctl/tpctl/scripts/env_raw.sh (hunk 8 above)
Comment (ofey404) [reply tipocket-ctl/tpctl/scripts/env_raw.sh]: Shorten it may make it harder to understand, especially for someone who don't familar with shell tools.
I've considered to finish the extraction of --namespace="" in a single sed line, but it would make this hard to read.

File: tipocket-ctl/tpctl/scripts/env_raw.sh (hunk 7 above)
Comment (ofey404) [reply tipocket-ctl/tpctl/scripts/env_raw.sh:1]: You are right.

File: tipocket-ctl/tpctl/scripts/env_raw.sh (hunk 8 above)
Comment (cosven) [reply tipocket-ctl/tpctl/scripts/env_raw.sh]: How about using `\`?

```
CLUSTER_NAMESPACE=$(argo get -n argo $DEPLOY_ID -o json | \
   jq -r '.spec.templates[-1].container.command[2]' | \
   grep -oP '\-namespace=".*?"' | \
   grep -o '".*"' | \
   sed -e 's/^"//' -e 's/"$//')
 

```

Comment (ofey404) [reply tipocket-ctl/tpctl/scripts/env_raw.sh]: Good

**Comment by cosven** [comment]
LGTM
